### PR TITLE
Feature/bulk download magres

### PR DESCRIPTION
**Pull request by Sathya-S3**
### New features in this pull request from feature/bulk_download_magres branch:
1. **Addition of new feature - bulk download of magres files with metadata in both JSON and CSV formats:**
    - Checkboxes added per search result, which can be checked manually by users. Suggested follow-up action is a click of "Download Selection (.zip)" button at the top of the search results to download the selection. This action defaults to the selection of latest version of each record for download (ignoring the dropdown menu choice). Selections made over multiple results pages are retained in memory until the user navigates to another tab (such as Home or Your Files, for example) without performing any downloads or the user initiates a new search or clicks the "Download All Results (.zip)" button.
    - Users can directly click the "Download All Results (.zip)" button at the top of the search results to download all the returned search results packed in a zip archive. By default, only the latest version of each record is downloaded in this case.
    - For both the above actions, a "Magres_metadata.json" file with comma-separated in-depth metadata for each downloaded magres file in JSON format and a "Magres_metadata.csv" file with human-readable metadata for the same are added to the archive.

2. **Modification to action triggered by "Download metadata (JSON)" button - now includes a JSON file clean-up prior to download:**
JSON metadata for a magres file is cleaned up to
    - remove redundant metadata fields,
    - restructure data fields like ORCID and version_history/last_version to a simplified form,
    - clean up date and calculation metadata information

before the JSON file is returned to the browser for download. This same action is performed for bulk download metadata as well.

### Modified files list + description of changes
1. **ccpncdb/server.py**
      - __init__ function now accepts an optional database object to use for CI tests (without the field, config is read from the config file as before), has new initialisation for the MetadataExport instance.
      - New functions added:
         **_unpack_file()_**: Unpacks file information from a dictionary.
         **_json_metadata_prepare()_**: Prepares JSON metadata for export.
         **_download_selection_json()_**: Downloads a selected JSON metadata file from the server.
         **_download_selection_zip()_**: Downloads selected Magres files as a ZIP archive with JSON and CSV metadata.

2. **main.py**
  Contains new flask routes for downloading magres files as a zip archive or for downloading just a single JSON file via POST requests.

3. **static/index.html**
  Load selectionservice.js and singleselectionservice.js scripts into HTML header when loading the CCPNC flask app.

4. **static/js/main.js**
  Load services defined in selectionservice.js and singleselectionservice.js scripts when loading the CCPNC flask app.

5. **static/templates/database_record.html**
    - Added checkbox per result,
    - Added spacing between new download buttons and search form box,
    - Added redirect to "Download metadata (JSON)" button to replace direct file retrieval with metadata clean up followed by retrieval.

6. **static/js/record.js**
    - Injecting the 'SelectionService' and 'SingleSelectionService' services into record.js.
    - Added functions for handling selection changes and JSON metadata retrieval for download. The function selectionChange() is used to compile the user selected records into a selectedItems list which assists in bulk file retrieval from the server. The jsonRetrieve() function performs the above action using a singleSelectJSON list to handle single JSON file downloads.

7. **static/templates/search_results.html**
  Added new bulk download buttons - "Download Selection (.zip)" and "Download All Results (.zip)".

8. **static/js/search.js**
    - Inject SelectionService into Search AngularJS controller,
    - Clear previous selections when initiating new search.

9. **static/js/searchresults.js**
    - Inject SelectionService and SingleSelectionService into Search Results AngularJS directive,
    - Introduced variable to store selection states of results,
    - New functions to trigger suitable action for json file download (download_json()), bulk download of user selection (download_selected_zip()), and bulk download of all results (download_all_zip()).

10. **tests/py/mdb_test.py**
  Added a CI test to verify the bulk download functionality. This test includes the following steps:
    **_testBulkDownload_**:
      - Adds test records to the database.
      - Retrieves records and reads their magresFileIDs.
      - Assigns filenames for bulk download.
      - Converts ObjectId fields to strings.
      - Compiles a list of files for bulk download.
      - Simulates a request to download the selected files.
      - Checks the response status code and content type.
      - Reads and verifies the contents of the ZIP archive.
      - Checks the contents of the JSON and CSV metadata files.
      - Verifies the contents of the magres files.

11. **static/css/stylesheet.css**
  Style changes to
    - resize results box,
    - override and adjust default Bulma checkbox styles to make them bigger.

12. **requirements.txt** and **setup.cfg**
  Change to pymongo version to be compatible with mongomock

### New files list + description
1. **ccpncdb/metadataexport.py**
  This file defines a MetadataExport class, which provides methods for cleaning and restructuring metadata in JSON format. The class includes functions to remove redundant fields, reorder keys, format dates, and extract calculation metadata. Functions summary:
    - **metadata_clearance()** - Cleans up metadata by removing redundant fields and restructuring specific entries.
    - **metadata_cleanup()** - Cleans up and reorders remaining metadata for a given JSON dataset.
    - **reorded_keys()** - Reorders the keys in a given JSON dictionary according to a predefined order.
    - **format_date()** - Formats a date string into a standardised form.
    - **calc_metadata_extract()** - Extracts calculation metadata from a JSON string and formats into a dictionary.

2. **static/js/selectionservice.js**
  A new AngularJS service to manage the selection and download of items in the selection page. This service includes the following functions:
    - clearAllSelections(): Clears the selectedItems and singleSelectJSON arrays,
    - clearSingleSelection(): Clears the singleSelectJSON array only (leaves selectedItems list as is, in case the "Download metadata (JSON)" button is clicked by mistake when making selections),
    - downloadSelectionZip(): Creates a ZIP archive of the selected items and initiates a download,
    - downloadSelectionJSON(): Creates a single JSON file of the selected record metadata and initiates a download.

3. **static/js/singleselectionservice.js**
  A new AngularJS service to manage the selection of individual items in the case of bulk downloading all results. This service includes the following function:
    - singleSelection(): Handles the selection and deselection of a single item. It adds the item to SelectionService.selectedItems list if selected, and removes it if deselected.
    - The above function is called in a loop from the download_all_zip() function of searchresults.js.

### CODE REVIEW

File: ccpncdb/metadataexport.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import datetime
+import json
+
+class MetadataExport:
```
Comment (jkshenton) [inline ccpncdb/metadataexport.py:4]: Are the methods for this class `@static_method` ?

File: ccpncdb/server.py [hunk 2]
```diff
@@ -69,12 +72,15 @@ def __init__(self, path=''):
             self._orcid_details, session)
         self._orcid = self._app.extensions['orcidlink']
 
-        self._config = Config(os.path.join(self._config_folder,
+        if db is None:
+            self._config = Config(os.path.join(self._config_folder,
```
Comment (jkshenton) [inline ccpncdb/server.py:76]: Do we need self._config anywhere else? If so, do we need to define this before the if/else? 

Similar Q for self._client, self._dbname and self._logger

File: ccpncdb/metadataexport.py (hunk 1 above)
Comment (Sathya-S3) [reply ccpncdb/metadataexport.py:4]: The methods of the class have been tagged as @staticmethod as they are grouped by the class but don't work on any instances of the class. They simply receive arguments and work on them, but grouped under the class for easier referencing within server.py. The _self_ reference in the methods' arguments have been removed.

File: ccpncdb/metadataexport.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            target_dict = {"Metadata Error": "Calculation metadata unavailable"}
+        else:
+            calc_string_unwrap = json.loads(calc_string)
+            for dict_elem in calc_string_unwrap:
+                if len(calc_string_unwrap[dict_elem][0]) == 1:
+                    target_dict[dict_elem] = calc_string_unwrap[dict_elem][0][0]
+                elif dict_elem == 'calc_pspot':
+                    final_elem = []
+                    for elem in calc_string_unwrap[dict_elem]:
+                        final_elem.append(' '.join(elem))
```
[the remark this answers is not in the document]
Comment (Sathya-S3) [reply ccpncdb/metadataexport.py]: calc_pspot has not been turned into a dictionary instead of a list.

File: ccpncdb/server.py (hunk 2 above)
Comment (Sathya-S3) [reply ccpncdb/server.py]: The mentioned instance variables have been moved out of the if-else block apart from self._db. Checked locally to not cause any errors. CI tests have passed.

Commit: [2dbc5ca](https://github.com/CCP-NC/ccpnc-database/pull/40/commits/2dbc5caf35adac466fae7a788607201babba9554)

**Review by jkshenton — APPROVED** [review]
Thanks for making the changes Sathya -- all looks good to me now

**Pull request merged**